File: uis/client/mobile_app/lib/services/esusu_service.dart
library;
import 'api_service.dart';
import 'dart:async';
import '../models/user.dart';
import 'user_service.dart';

class EsusuService {
    /// Get user details from local storage
    Future<User?> getLocalUser() async {
      final userService = UserService();
      return await userService.getStoredUserII();
    }
  static EsusuService? _instance;
  static EsusuService get instance => _instance ??= EsusuService._();

  final ApiService _apiClient = ApiService();
  EsusuService._();

  // Cache to store group data with members
  final Map<String, EsusuGroup> _groupCache = {};

  Future<List<EsusuGroup>> getGroups({
    EsusuGroupStatus? status,
    bool? myGroups,
    int page = 1,
    int pageSize = 20,
  }) async {
    try {
      final queryParams = <String, dynamic>{
        // if (status != null) 'status': status.value,
        if (myGroups == true) 'my_groups': true,
        'page': page,
        'limit': pageSize,
      };

      final response = await _apiClient.get(
        '/esusu/api/v1/esusu/groups',
        queryParameters: queryParams,
      );

      // Handle different response structures
      List<dynamic> itemsRaw;
      if (response.data is List) {
        itemsRaw = response.data as List;
      } else if (response.data is Map) {
        final dataMap = response.data as Map<String, dynamic>;
        itemsRaw = (dataMap['groups'] ?? dataMap['data'] ?? []) as List;
      } else {
        itemsRaw = [];
      }

      final List<EsusuGroup> groups = [];
      for (var i = 0; i < itemsRaw.length; i++) {
        try {
          final item = itemsRaw[i];
          if (item is Map<String, dynamic>) {
            final group = EsusuGroup.fromJson(item);
            groups.add(group);
            _groupCache[group.id] = group;
          }
        } catch (itemError) {
          // Skip malformed items
        }
      }

      return groups;
    } catch (e) {
      return [];
    }
  }

  Future<EsusuGroup> getGroup(String groupId) async {
    final response = await _apiClient.get('/esusu/api/v1/esusu/groups/$groupId');
    final group = EsusuGroup.fromJson(response.data as Map<String, dynamic>);
    _groupCache[groupId] = group;
    return group;
  }

  Future<EsusuGroup> createGroup({
    required String name,
    String? description,
    required double contributionAmount,
    required ContributionFrequency frequency,
    required int maxMembers,
    required DateTime startDate,
    String? creatorId,
    String? creatorName,
  }) async {
    User? user;
    if (creatorId == null || creatorName == null) {
      user = await getLocalUser();
    }
    final payload = {
      'name': name,
      'description': description,
      'contribution_amount': contributionAmount,
      'frequency': frequency.value,
      'max_members': maxMembers,
      'start_date': startDate.toUtc().toIso8601String(),
      'created_by': creatorId ?? user?.id,
      'creator_name': creatorName ?? user?.fullName,
    };
    final response = await _apiClient.post('/esusu/api/v1/esusu/groups', data: payload);
    final group = EsusuGroup.fromJson(response.data as Map<String, dynamic>);
    _groupCache[group.id] = group;
    return group;
  }

  Future<EsusuMember> joinGroup(
    String groupId) async {
    User? user;
      user = await getLocalUser();
      if (user == null || user.id.isEmpty || user.fullName.isEmpty || user.phoneNumber == null || user.phoneNumber!.isEmpty) {

        throw Exception('No valid local user found, $user');
      }

    final payload = {
      'user_id':  user.id,
      'name':  user.fullName,
      'phone': user.phoneNumber,

    };
    final response = await _apiClient.post(
      '/esusu/api/v1/esusu/groups/$groupId/join',
      data: payload,
    );
    _groupCache.remove(groupId);
    return EsusuMember.fromJson(response.data as Map<String, dynamic>);
  }

  /// Get members from group data
  /// Since members are included in the group response, we just fetch the group
  Future<List<EsusuMember>> getGroupMembers(String groupId) async {
    try {
      final group = await getGroup(groupId);
      return group.members;
    } catch (e) {
      return [];
    }
  }

  Future<EsusuContribution> makeContribution(
    String groupId, {
    String? memberId,
    double? amount,
    required DateTime expectedDate,
    required String paymentMethod,
  }) async {
    User? user;
      user = await getLocalUser();
      if (user == null || user.id.isEmpty) {
        throw Exception('No valid local user found, $user');
      }
    final payload = {
      'member_id':  user.id,
      'amount': amount ?? 0.0,
      'expected_date': expectedDate.toUtc().toIso8601String(),
      'payment_method': paymentMethod,
    };
    final response = await _apiClient.post(
      '/esusu/api/v1/esusu/groups/$groupId/contributions',
      data: payload,
    );
    return EsusuContribution.fromJson(response.data as Map<String, dynamic>);
  }

  Future<List<EsusuContribution>> getContributions(String groupId, {int? round}) async {
    try {
      final queryParams = <String, dynamic>{};
      if (round != null) queryParams['round'] = round;

      final response = await _apiClient.get(
        '/esusu/api/v1/esusu/groups/$groupId/contributions',
        queryParameters: queryParams,
      );

      List<dynamic> itemsRaw;
      if (response.data is List) {
        itemsRaw = response.data as List;
      } else if (response.data is Map) {
        final dataMap = response.data as Map<String, dynamic>;
        itemsRaw = (dataMap['contributions'] ?? dataMap['data'] ?? []) as List;
      } else {
        itemsRaw = [];
      }

      final List<EsusuContribution> contributions = [];
      for (var item in itemsRaw) {
        if (item is Map<String, dynamic>) {
          contributions.add(EsusuContribution.fromJson(item));
        }
      }

      return contributions;
    } catch (e) {
      return [];
    }
  }

  Future<List<EsusuPayout>> getPayouts(String groupId) async {
    try {
      final response = await _apiClient.get(
        '/esusu/api/v1/esusu/groups/$groupId/payouts',
      );

      List<dynamic> itemsRaw;
      if (response.data is List) {
        itemsRaw = response.data as List;
      } else if (response.data is Map) {
        final dataMap = response.data as Map<String, dynamic>;
        itemsRaw = (dataMap['payouts'] ?? dataMap['data'] ?? []) as List;
      } else {
        itemsRaw = [];
      }

      final List<EsusuPayout> payouts = [];
      for (var item in itemsRaw) {
        if (item is Map<String, dynamic>) {
          payouts.add(EsusuPayout.fromJson(item));
        }
      }

      return payouts;
    } catch (e) {
      return [];
    }
  }

  Future<EsusuPayout> processPayout(String groupId) async {
    final response = await _apiClient.post('/esusu/api/v1/esusu/groups/$groupId/payout');
    return EsusuPayout.fromJson(response.data as Map<String, dynamic>);
  }

  /// Clear cache for a specific group or all groups
  void clearCache([String? groupId]) {
    if (groupId != null) {
      _groupCache.remove(groupId);
    } else {
      _groupCache.clear();
    }
  }
}

/// Esusu group status types
enum EsusuGroupStatus {
  pending,
  active,
  completed,
  cancelled,
}

extension EsusuGroupStatusExtension on EsusuGroupStatus {
  String get value {
    switch (this) {
      case EsusuGroupStatus.pending:
        return 'pending';
      case EsusuGroupStatus.active:
        return 'active';
      case EsusuGroupStatus.completed:
        return 'completed';
      case EsusuGroupStatus.cancelled:
        return 'cancelled';
    }
  }

  static EsusuGroupStatus fromString(String value) {
    switch (value.toLowerCase()) {
      case 'pending':
        return EsusuGroupStatus.pending;
      case 'active':
        return EsusuGroupStatus.active;
      case 'completed':
        return EsusuGroupStatus.completed;
      case 'cancelled':
        return EsusuGroupStatus.cancelled;
      default:
        return EsusuGroupStatus.pending;
    }
  }
}

/// Esusu member status types
enum EsusuMemberStatus {
  pending,
  active,
  defaulted,
  completed,
}

extension EsusuMemberStatusExtension on EsusuMemberStatus {
  String get value {
    switch (this) {
      case EsusuMemberStatus.pending:
        return 'pending';
      case EsusuMemberStatus.active:
        return 'active';
      case EsusuMemberStatus.defaulted:
        return 'defaulted';
      case EsusuMemberStatus.completed:
        return 'completed';
    }
  }

  static EsusuMemberStatus fromString(String value) {
    switch (value.toLowerCase()) {
      case 'pending':
        return EsusuMemberStatus.pending;
      case 'active':
        return EsusuMemberStatus.active;
      case 'defaulted':
        return EsusuMemberStatus.defaulted;
      case 'completed':
        return EsusuMemberStatus.completed;
      default:
        return EsusuMemberStatus.pending;
    }
  }
}

/// Contribution frequency types
enum ContributionFrequency {
  daily,
  weekly,
  biweekly,
  monthly,
}

extension ContributionFrequencyExtension on ContributionFrequency {
  String get value {
    switch (this) {
      case ContributionFrequency.daily:
        return 'daily';
      case ContributionFrequency.weekly:
        return 'weekly';
      case ContributionFrequency.biweekly:
        return 'biweekly';
      case ContributionFrequency.monthly:
        return 'monthly';
    }
  }

  static ContributionFrequency fromString(String value) {
    switch (value.toLowerCase()) {
      case 'daily':
        return ContributionFrequency.daily;
      case 'weekly':
        return ContributionFrequency.weekly;
      case 'biweekly':
        return ContributionFrequency.biweekly;
      case 'monthly':
        return ContributionFrequency.monthly;
      default:
        return ContributionFrequency.monthly;
    }
  }
}

/// Esusu group model
class EsusuGroup {
  final String id;
  final String name;
  final String? description;
  final EsusuGroupStatus status;
  final double contributionAmount;
  final ContributionFrequency frequency;
  final int maxMembers;
  final int currentMembers;
  final int currentRound;
  final int totalRounds;
  final DateTime startDate;
  final DateTime? endDate;
  final String creatorId;
  final String creatorName;
  final DateTime createdAt;
  final List<EsusuMember> members; // Members are included in group data

  EsusuGroup({
    required this.id,
    required this.name,
    this.description,
    required this.status,
    required this.contributionAmount,
    required this.frequency,
    required this.maxMembers,
    required this.currentMembers,
    required this.currentRound,
    required this.totalRounds,
    required this.startDate,
    this.endDate,
    required this.creatorId,
    required this.creatorName,
    required this.createdAt,
    required this.members,
  });

  factory EsusuGroup.fromJson(Map<String, dynamic> json) {
    final List<EsusuMember> membersList = [];
    if (json['members'] != null && json['members'] is List) {
      final membersData = json['members'] as List;
      for (var i = 0; i < membersData.length; i++) {
        try {
          final memberItem = membersData[i];
          if (memberItem is Map<String, dynamic>) {
            membersList.add(EsusuMember.fromJson(memberItem));
          }
        } catch (_) {
          // Skip malformed member
        }
      }
    }

    return EsusuGroup(
      id: json['id']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      description: json['description']?.toString(),
      status: EsusuGroupStatusExtension.fromString(
        (json['status'] ?? 'pending').toString(),
      ),
      contributionAmount: (json['contribution_amount'] ?? json['contributionAmount'] ?? 0).toDouble(),
      frequency: ContributionFrequencyExtension.fromString(
        (json['frequency'] ?? 'monthly').toString(),
      ),
      maxMembers: (json['max_members'] ?? json['maxMembers'] ?? 0) as int,
      currentMembers: membersList.length,
      currentRound: (json['current_cycle'] ?? json['currentRound'] ?? 0) as int,
      totalRounds: (json['total_cycles'] ?? json['totalRounds'] ?? 0) as int,
      startDate: json['start_date'] != null
          ? DateTime.parse(json['start_date'].toString())
          : DateTime.now(),
      endDate: json['end_date'] != null
          ? DateTime.parse(json['end_date'].toString())
          : null,
      creatorId: json['created_by']?.toString() ?? json['creator_id']?.toString() ?? '',
      // creator_name takes priority over admin_id (which is an ID, not a display name)
      creatorName: json['creator_name']?.toString() ?? json['admin_name']?.toString() ?? 'Unknown',
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'].toString())
          : DateTime.now(),
      members: membersList,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      if (description != null) 'description': description,
      'status': status.value,
      'contribution_amount': contributionAmount,
      'frequency': frequency.value,
      'max_members': maxMembers,
      'current_members': currentMembers,
      'current_round': currentRound,
      'total_rounds': totalRounds,
      'start_date': startDate.toIso8601String(),
      if (endDate != null) 'end_date': endDate!.toIso8601String(),
      'creator_id': creatorId,
      'creator_name': creatorName,
      'created_at': createdAt.toIso8601String(),
      'members': members.map((m) => m.toJson()).toList(),
    };
  }
}

/// Esusu member model
class EsusuMember {
  final String id;
  final String groupId;
  final String userId;
  final String fullName;
  final String phone;
  final EsusuMemberStatus status;
  final int position;
  final double totalContributed;
  final double totalReceived;
  final bool hasReceivedPayout;
  final DateTime joinedAt;

  EsusuMember({
    required this.id,
    required this.groupId,
    required this.userId,
    required this.fullName,
    required this.phone,
    required this.status,
    required this.position,
    required this.totalContributed,
    required this.totalReceived,
    required this.hasReceivedPayout,
    required this.joinedAt,
  });

  factory EsusuMember.fromJson(Map<String, dynamic> json) {
    return EsusuMember(
      id: json['id']?.toString() ?? '',
      groupId: json['group_id']?.toString() ?? '',
      userId: json['user_id']?.toString() ?? '',
      fullName: json['full_name']?.toString() ?? json['name']?.toString() ?? 'Unknown',
      phone: json['phone']?.toString() ?? '',
      status: EsusuMemberStatusExtension.fromString(
        (json['status'] ?? 'pending').toString(),
      ),
      position: (json['position'] ?? 0) as int,
      totalContributed: (json['total_contributed'] ?? 0).toDouble(),
      totalReceived: (json['total_received'] ?? 0).toDouble(),
      hasReceivedPayout: json['has_received_payout'] == true,
      joinedAt: json['joined_at'] != null
          ? DateTime.parse(json['joined_at'].toString())
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'group_id': groupId,
      'user_id': userId,
      'full_name': fullName,
      'phone': phone,
      'status': status.value,
      'position': position,
      'total_contributed': totalContributed,
      'total_received': totalReceived,
      'has_received_payout': hasReceivedPayout,
      'joined_at': joinedAt.toIso8601String(),
    };
  }
}

/// Esusu contribution model
class EsusuContribution {
  final String id;
  final String groupId;
  final String memberId;
  final String memberName;
  final int round;
  final double amount;
  final DateTime contributedAt;
  final String? transactionRef;

  EsusuContribution({
    required this.id,
    required this.groupId,
    required this.memberId,
    required this.memberName,
    required this.round,
    required this.amount,
    required this.contributedAt,
    this.transactionRef,
  });

  factory EsusuContribution.fromJson(Map<String, dynamic> json) {
    return EsusuContribution(
      id: json['id']?.toString() ?? '',
      groupId: json['group_id']?.toString() ?? '',
      memberId: json['member_id']?.toString() ?? '',
      memberName: json['member_name']?.toString() ?? '',
      round: (json['round'] ?? 0) as int,
      amount: (json['amount'] ?? 0).toDouble(),
      contributedAt: json['contributed_at'] != null
          ? DateTime.parse(json['contributed_at'].toString())
          : DateTime.now(),
      transactionRef: json['transaction_ref']?.toString(),
    );
  }
}

/// Esusu payout model
class EsusuPayout {
  final String id;
  final String groupId;
  final String memberId;
  final String memberName;
  final int round;
  final double amount;
  final DateTime paidAt;
  final String? transactionRef;

  EsusuPayout({
    required this.id,
    required this.groupId,
    required this.memberId,
    required this.memberName,
    required this.round,
    required this.amount,
    required this.paidAt,
    this.transactionRef,
  });

  factory EsusuPayout.fromJson(Map<String, dynamic> json) {
    return EsusuPayout(
      id: json['id']?.toString() ?? '',
      groupId: json['group_id']?.toString() ?? '',
      memberId: json['member_id']?.toString() ?? '',
      memberName: json['member_name']?.toString() ?? '',
      round: (json['round'] ?? 0) as int,
      amount: (json['amount'] ?? 0).toDouble(),
      paidAt: json['paid_at'] != null
          ? DateTime.parse(json['paid_at'].toString())
          : DateTime.now(),
      transactionRef: json['transaction_ref']?.toString(),
    );
  }
}

/// Esusu stats model
class EsusuStats {
  final int totalGroups;
  final int activeGroups;
  final int totalMembers;
  final double totalContributions;
  final double totalPayouts;
  final int myGroups;
  final double myContributions;
  final double myPayouts;

  EsusuStats({
    required this.totalGroups,
    required this.activeGroups,
    required this.totalMembers,
    required this.totalContributions,
    required this.totalPayouts,
    required this.myGroups,
    required this.myContributions,
    required this.myPayouts,
  });

  factory EsusuStats.fromJson(Map<String, dynamic> json) {
    return EsusuStats(
      totalGroups: (json['total_groups'] ?? 0) as int,
      activeGroups: (json['active_groups'] ?? 0) as int,
      totalMembers: (json['total_members'] ?? 0) as int,
      totalContributions: (json['total_contributions'] ?? 0).toDouble(),
      totalPayouts: (json['total_payouts'] ?? 0).toDouble(),
      myGroups: (json['my_groups'] ?? 0) as int,
      myContributions: (json['my_contributions'] ?? 0).toDouble(),
      myPayouts: (json['my_payouts'] ?? 0).toDouble(),
    );
  }
}

File: uis/client/web2/src/services/esusu_service.ts
import type { EsusuGroup, EsusuGroupStatus as EsusuGroupStatusType, EsusuMember, EsusuStats } from '../models/esusu';
import { apiService } from './api_service';

interface EsusuContribution {
  id: string;
  memberId: string;
  memberName: string;
  amount: number;
  round: number;
  date: string;
  paymentMethod: string;
}

interface ContributionRequest {
  memberId: string;
  amount: number;
  expectedDate: Date;
  paymentMethod: string;
}

class EsusuService {
  private groupCache: Map<string, EsusuGroup> = new Map();

  async getStats(): Promise<EsusuStats> {
    const response = await apiService.get('/esusu/api/v1/esusu/stats');
    return response.data as EsusuStats;
  }

  async getGroups(myGroups = false, status?: EsusuGroupStatusType, page = 1, pageSize = 20): Promise<EsusuGroup[]> {
    const params: Record<string, unknown> = {
      page,
      limit: pageSize,
    };
    
    if (myGroups) {
      params.my_groups = true;
    }
    if (status) {
      params.status = status.toLowerCase();
    }

    const response = await apiService.get('/esusu/api/v1/esusu/groups', params);
    
    // Handle different response structures like mobile does
    let itemsRaw: unknown[];
    if (Array.isArray(response.data)) {
      itemsRaw = response.data;
    } else if (response.data && typeof response.data === 'object') {
      const dataMap = response.data as Record<string, unknown>;
      itemsRaw = (dataMap.groups || dataMap.data || []) as unknown[];
    } else {
      itemsRaw = [];
    }
    
    const groups: EsusuGroup[] = [];
    for (const item of itemsRaw) {
      if (item && typeof item === 'object') {
        const group = this.parseGroup(item as Record<string, unknown>);
        groups.push(group);
        // Cache the group data including members
        this.groupCache.set(group.id, group);
      }
    }
    
    return groups;
  }

  async getGroupById(id: string): Promise<EsusuGroup> {
    const response = await apiService.get(`/esusu/api/v1/esusu/groups/${id}`);
    const group = this.parseGroup(response.data as Record<string, unknown>);
    // Cache the group data including members
    this.groupCache.set(id, group);
    return group;
  }

  async getGroupMembers(groupId: string): Promise<EsusuMember[]> {
    // Members are included in the group response, so fetch the group
    const group = await this.getGroupById(groupId);
    return group.members || [];
  }

  async createGroup(data: {
    name: string;
    description?: string;
    contributionAmount: number;
    frequency: 'daily' | 'weekly' | 'monthly';
    maxMembers: number;
    startDate: Date;
  }): Promise<EsusuGroup> {
    const payload = {
      name: data.name,
      description: data.description,
      contribution_amount: data.contributionAmount,
      frequency: data.frequency,
      max_members: data.maxMembers,
      start_date: data.startDate.toISOString(),
    };
    
    const response = await apiService.post('/esusu/api/v1/esusu/groups', payload);
    const group = this.parseGroup(response.data as Record<string, unknown>);
    this.groupCache.set(group.id, group);
    return group;
  }

  async joinGroup(groupId: string): Promise<EsusuMember> {
    const response = await apiService.post(`/esusu/api/v1/esusu/groups/${groupId}/join`, {});
    // Clear cache so it reloads with new member
    this.groupCache.delete(groupId);
    return this.parseMember(response.data as Record<string, unknown>);
  }

  async leaveGroup(groupId: string): Promise<void> {
    await apiService.post(`/esusu/api/v1/esusu/groups/${groupId}/leave`, {});
    this.groupCache.delete(groupId);
  }

  async getContributions(groupId: string, round?: number): Promise<EsusuContribution[]> {
    const params: Record<string, unknown> = {};
    if (round !== undefined) {
      params.round = round;
    }
    
    const response = await apiService.get(`/esusu/api/v1/esusu/groups/${groupId}/contributions`, params);
    
    // Handle different response structures
    let itemsRaw: unknown[];
    if (Array.isArray(response.data)) {
      itemsRaw = response.data;
    } else if (response.data && typeof response.data === 'object') {
      const dataMap = response.data as Record<string, unknown>;
      itemsRaw = (dataMap.contributions || dataMap.data || []) as unknown[];
    } else {
      itemsRaw = [];
    }
    
    return itemsRaw.map(item => this.parseContribution(item as Record<string, unknown>));
  }

  async makeContribution(groupId: string, data: ContributionRequest): Promise<EsusuContribution> {
    const payload = {
      member_id: data.memberId,
      amount: data.amount,
      expected_date: data.expectedDate.toISOString(),
      payment_method: data.paymentMethod,
    };
    
    const response = await apiService.post(`/esusu/api/v1/esusu/groups/${groupId}/contributions`, payload);
    return this.parseContribution(response.data as Record<string, unknown>);
  }

  clearCache(groupId?: string): void {
    if (groupId) {
      this.groupCache.delete(groupId);
    } else {
      this.groupCache.clear();
    }
  }

  private parseGroup(data: Record<string, unknown>): EsusuGroup {
    const membersRaw = data.members as unknown[] | undefined;
    const members: EsusuMember[] = [];
    if (Array.isArray(membersRaw)) {
      for (const memberItem of membersRaw) {
        if (memberItem && typeof memberItem === 'object') {
          members.push(this.parseMember(memberItem as Record<string, unknown>));
        }
      }
    }

    return {
      id: String(data.id || ''),
      name: String(data.name || ''),
      description: String(data.description || ''),
      contributionAmount: Number(data.contribution_amount || data.contributionAmount || 0),
      frequency: (data.frequency as 'daily' | 'weekly' | 'monthly') || 'monthly',
      totalMembers: Number(data.max_members || data.totalMembers || 0),
      currentMembers: Number(data.current_members || data.currentMembers || members.length),
      cycleNumber: Number(data.current_cycle || data.cycle_number || data.cycleNumber || 0),
      status: String(data.status || 'pending'),
      // creator_name takes priority; admin_name as fallback
      creatorName: String(data.creator_name || data.admin_name || ''),
      members,
    };
  }

  private parseMember(data: Record<string, unknown>): EsusuMember {
    // Handle name field - API returns 'name', fallback to 'Unknown Member' if empty
    const name = data.name || data.full_name || data.fullName || '';
    const fullName = String(name).trim() || 'Unknown Member';
    
    return {
      id: String(data.id || ''),
      fullName,
      position: Number(data.position || data.payout_position || 0),
      hasReceivedPayout: Boolean(data.has_received_payout || data.hasReceivedPayout || false),
    };
  }

  private parseContribution(data: Record<string, unknown>): EsusuContribution {
    return {
      id: String(data.id || ''),
      memberId: String(data.member_id || data.memberId || ''),
      memberName: String(data.member_name || data.memberName || ''),
      amount: Number(data.amount || 0),
      round: Number(data.round || data.cycle_number || 1),
      date: String(data.contribution_date || data.date || new Date().toISOString()),
      paymentMethod: String(data.payment_method || data.paymentMethod || ''),
    };
  }
}

export const esusuService = new EsusuService();

